salt/cli/support/collector.py

```diff
@@ -0,0 +1,497 @@
+# coding=utf-8
+from __future__ import absolute_import, print_function, unicode_literals
+import os
+import sys
+import copy
+import yaml
+import json
+import logging
+import tarfile
+import time
+import salt.ext.six as six
+
+if six.PY2:
+    import exceptions
+else:
+    import builtins as exceptions
+    from io import IOBase as file
+
+from io import BytesIO
+
+sys.modules['pkg_resources'] = None
+
+import salt.utils.stringutils
+import salt.utils.parsers
+import salt.utils.verify
+import salt.utils.platform
+import salt.utils.process
+import salt.exceptions
+import salt.defaults.exitcodes
+import salt.cli.caller
+import salt.cli.support
+import salt.cli.support.console
+import salt.cli.support.intfunc
+import salt.cli.support.localrunner
+import salt.output.table_out
+import salt.runner
+import salt.utils.files
+
+
+salt.output.table_out.__opts__ = {}
+log = logging.getLogger(__name__)
+
+
+class SupportDataCollector(object):
+    '''
+    Data collector. It behaves just like another outputter,
+    except it grabs the data to the archive files.
+    '''
+    def __init__(self, name, output):
+        '''
+        constructor of the data collector
+        :param name:
+        :param path:
+        :param format:
+        '''
+        self.archive_path = name
+        self.__default_outputter = output
+        self.__format = format
+        self.__arch = None
+        self.__current_section = None
+        self.__current_section_name = None
+        self.__default_root = time.strftime('%Y.%m.%d-%H.%M.%S-snapshot')
+        self.out = salt.cli.support.console.MessagesOutput()
+
+    def open(self):
+        '''
+        Opens archive.
+        :return:
+        '''
+        if self.__arch is not None:
+            raise salt.exceptions.SaltException('Archive already opened.')
+        self.__arch = tarfile.TarFile.bz2open(self.archive_path, 'w')
+
+    def close(self):
+        '''
+        Closes the archive.
+        :return:
+        '''
+        if self.__arch is None:
+            raise salt.exceptions.SaltException('Archive already closed')
+        self._flush_content()
+        self.__arch.close()
+        self.__arch = None
+
+    def _flush_content(self):
```

Comment:
Here's another function that CodeClimate has marked as being difficult to read. Can we break this up into something easier to understand?